packages/gatsby/src/utils/worker/render-html.ts

```diff
@@ -1,50 +1,315 @@
+  const rawPageData = await fs.readFile(filePath, `utf-8`)
+
+  return JSON.parse(rawPageData)
+}
+
+async function readWebpackStats(publicDir: string): any {
+  const filePath = join(publicDir, `webpack.stats.json`)
+  const rawPageData = await fs.readFile(filePath, `utf-8`)
+
+  return JSON.parse(rawPageData)
+}
+
+interface IScriptsAndStyles {
+  scripts: Array<any>
+  styles: Array<any>
+  reversedStyles: Array<any>
+  reversedScripts: Array<any>
+}
+
+async function getScriptsAndStylesForTemplate(
+  componentChunkName
+): Promise<IScriptsAndStyles> {
+  // Create paths to scripts
+  let scriptsAndStyles = flatten(
+    [`app`, componentChunkName].map(s => {
+      const fetchKey = `assetsByChunkName[${s}]`
+
+      let chunks = get(webpackStats, fetchKey)
+      const namedChunkGroups = get(webpackStats, `namedChunkGroups`)
+
+      if (!chunks) {
+        return null
+      }
+
+      chunks = chunks.map(chunk => {
+        if (chunk === `/`) {
+          return null
+        }
+        return { rel: `preload`, name: chunk }
+      })
+
+      namedChunkGroups[s].assets.forEach(asset =>
+        chunks.push({ rel: `preload`, name: asset })
+      )
+
+      const childAssets = namedChunkGroups[s].childAssets
+      for (const rel in childAssets) {
+        chunks = concat(
+          chunks,
+          childAssets[rel].map(chunk => {
+            return { rel, name: chunk }
+          })
+        )
+      }
+
+      return chunks
+    })
+  )
+    .filter(s => isObject(s))
+    .sort((s1, s2) => (s1.rel == `preload` ? -1 : 1)) // given priority to preload
+
+  scriptsAndStyles = uniqBy(scriptsAndStyles, item => item.name)
+
+  const scripts = scriptsAndStyles.filter(
+    script => script.name && script.name.endsWith(`.js`)
+  )
+  const styles = scriptsAndStyles.filter(
+    style => style.name && style.name.endsWith(`.css`)
+  )
+
+  return {
+    scripts,
+    styles,
+    reversedStyles: await Promise.all(
+      styles
+        .slice(0)
+        .reverse()
+        .map(async style => {
+          if (style.rel !== `prefetch`) {
+            const memoizedInlineCss = inlineCssCache.get(style.name)
+            if (memoizedInlineCss) {
+              return {
+                ...style,
+                content: memoizedInlineCss,
+              }
+            }
+
+            let getInlineCssPromise = inFlightInlineCssPromise.get(style.name)
+            if (!getInlineCssPromise) {
+              getInlineCssPromise = fs
+                .readFile(join(process.cwd(), `public`, style.name), `utf-8`)
+                .then(content => {
+                  inlineCssCache.set(style.name, content)
+                  inFlightInlineCssPromise.delete(style.name)
+
+                  return content
+                })
+
+              inFlightInlineCssPromise.set(style.name, getInlineCssPromise)
+            }
+
+            return {
+              ...style,
+              content: await getInlineCssPromise,
+            }
```

Comment:
Nit: do we need this `inlineCssCache` cache? Maybe we can just keep `inlineCssPromiseCache` to simplify this a bit? 

```suggestion
          if (style.rel !== `prefetch`) {
            let getInlineCssPromise = inFlightInlineCssPromise.get(style.name)
            if (!getInlineCssPromise) {
              getInlineCssPromise = fs
                .readFile(join(process.cwd(), `public`, style.name), `utf-8`)

              inFlightInlineCssPromise.set(style.name, getInlineCssPromise)
            }

            return {
              ...style,
              content: await getInlineCssPromise,
            }
```

We still preserve `inlineCssCache` in memory so not a big difference. Or am I missing something?

Reply:
No, you are right. This was just poor attempt at avoiding promise usage when not needed, but given that this would be used at most once per template there's no measurable diff there to gain and simplifying things is for the best

Reply:
done in https://github.com/gatsbyjs/gatsby/pull/29396/commits/f9016bb97cdca446e1f79cf300c86b82319c73a1